flask-version/metagenomongo/app.py

```diff
@@ -26,36 +27,58 @@ def allowed_file(filename):
     return '.' in filename and \
            filename.rsplit('.', 1)[-1].lower() in ALLOWED_EXTENSIONS
 
+@app.route('/change', methods=['POST'])
+def change():
+    form_data = request.form
+    data_list = []
+    tmp = []
+    count = 0
+    for key, value in form_data.items():
+        tmp.append(value)
+        count += 1
+        if count == 86:
+            data_list.append(tmp)
+            tmp = []
+            count = 0
+    # return redirect("/")
+    results = []
+    values = {"default": 0}
+    data = pd.DataFrame(data_list, columns=fields)
+    data_validation.validation_all( fields, options, results, data)
+    return render_template('index_with_table.html', \
+                    tables=[data.to_html(classes='data', header="true")], fields=fields, results=results, values=values, df=data)
+
+
 @app.route('/save', methods=['GET', 'POST'])
 def save():
-    csv_data = request.form['csv_data']
-    df = pd.read_html(csv_data, index_col=None)[0]
+    form_data = request.form
+    data_list = []
+    tmp = []
+    count = 0
+    for key, value in form_data.items():
+        tmp.append(value)
+        count += 1
+        if count == 86:
+            data_list.append(tmp)
+            tmp = []
+            count = 0
+    df= pd.DataFrame(data_list, columns=fields)
     df = df.loc[:, ~df.columns.str.contains('^Unnamed')]
     output = io.StringIO()
     df.to_csv(output, index=False)
     mem = io.BytesIO()
     mem.write(output.getvalue().encode('utf-8'))
     mem.seek(0)
     email.send_email() 
+    current_time = datetime.datetime.now()
+    file_name= form_data["user_name"] + current_time.strftime('_%Y-%m-%d-%H:%M:%S') +".csv"
     return send_file(mem, mimetype='text/csv', \
-                     as_attachment=True, download_name='metamongo_file.csv')
-
-@app.route('/correct', methods=['GET', 'POST'])
-def correct():
-    results = []
-    values = {"default": 0}
-    values = request.form
-    csv_data = request.form['csv_data']
-    data = pd.read_html(csv_data, index_col=None)[0]
-    data = data.loc[:, ~data.columns.str.contains('^Unnamed')]
-    data = data.fillna("")
-    data = data.astype(str)
-    data_validation.validation_all(fields, options, results, data)
-    return render_template('index.html', tables=[data.to_html(classes='data', header="true")], fields=fields, results=results, values=values)
+                     as_attachment=True, download_name=file_name)
 
 @app.route('/', methods=['GET', 'POST'])
 def index():
     # row, column_name, error type
+    global fields
```

Comment:
@iwmstjp Definitely review this and implement the necessary change. Since we explicitly want to allow the application to run in a multi-threaded environment by several users concurrently, we need it to be thread-safe. Global variables are not considered good programming practice in Python anyway, so avoid them whenever possible even when multi-threading is not a major concern.